react/AppIcon/test/AppIcon.spec.js

```diff
@@ -7,39 +7,128 @@ import { shallow } from 'enzyme'
-    ).getElement()
+    const wrapper = shallow(
+      <AppIcon
+        app={app}
+        domain={domain}
+        fetchIcon={successFetchIcon}
+        secure={secure}
+      />
+    )
+
+    const component = wrapper.getElement()
     expect(component).toMatchSnapshot()
+    expect(successFetchIcon).toHaveBeenCalledWith(app, domain, secure)
+    expect(console.error).toHaveBeenCalledTimes(0)
   })
 
-  it('renders app icon after fetching', async () => {
-    const successFetchPromise = successFetchIcon()
+  it(`renders correctly`, async done => {
     const wrapper = shallow(
-      <AppIcon app={app} fetchIcon={() => successFetchPromise} />
+      <AppIcon
+        app={app}
+        domain={domain}
+        fetchIcon={successFetchIcon}
+        onReady={() => {
+          wrapper.update()
+          const component = wrapper.getElement()
+          expect(component).toMatchSnapshot()
+          expect(successFetchIcon).toHaveBeenCalledWith(app, domain, secure)
+          expect(console.error).toHaveBeenCalledTimes(0)
+          done()
+        }}
+        secure={secure}
+      />
+    )
+  })
+
+  it(`renders default when fetch error occurs`, async done => {
+    const wrapper = shallow(
+      <AppIcon
+        app={app}
+        domain={domain}
+        fetchIcon={failureFetchIcon}
+        onReady={() => {
+          wrapper.update()
+          const component = wrapper.getElement()
+          expect(component).toMatchSnapshot()
+          expect(failureFetchIcon).toHaveBeenCalledWith(app, domain, secure)
+          expect(console.error).toHaveBeenCalledTimes(0)
+          done()
+        }}
+        secure={secure}
+      />
+    )
+  })
+
+  it(`uses Preloader.preload when no fetchIcon method is provided`, async done => {
+    jest.mock('../Preloader')
+    const AppIcon = require('../').default
+    const Preloader = require('../Preloader')
+    const wrapper = shallow(
+      <AppIcon
+        app={app}
+        domain={domain}
+        onReady={() => {
+          wrapper.update()
+          const component = wrapper.getElement()
+          expect(component).toMatchSnapshot()
+          expect(Preloader.preload).toHaveBeenCalledWith(app, domain, secure)
+          expect(console.error).toHaveBeenCalledTimes(0)
+          done()
+        }}
+        secure={secure}
+      />
+    )
+  })
+
+  it(`renders immediately when icon is alredy preloaded`, async done => {
+    jest.mock('../Preloader')
+    const AppIcon = require('../').default
+    const Preloader = require('../Preloader')
+
+    shallow(
+      <AppIcon
+        app={app}
+        domain={domain}
+        onReady={() => {
+          const wrapper = shallow(
+            <AppIcon app={app} domain={domain} secure={secure} />
+          )
+          const component = wrapper.getElement()
+          expect(component).toMatchSnapshot()
+          expect(Preloader.getPreloaded).toHaveBeenCalledWith(
+            app,
+            domain,
+            secure
+          )
```

Comment:
I am not sure if this should be tested as the preloader is an implementation detail.

Reply:
We target to use the image preloading as default behaviour, now that icons are accessible without permissions, so I think it is important to test it.